include/SampleBuffer.h

```diff
@@ -51,17 +52,52 @@ class LMMS_EXPORT SampleBuffer
 	using reverse_iterator = std::vector<sampleFrame>::reverse_iterator;
 	using const_reverse_iterator = std::vector<sampleFrame>::const_reverse_iterator;
 
-	SampleBuffer() = default;
-	explicit SampleBuffer(const QString& audioFile);
-	SampleBuffer(const QString& base64, int sampleRate);
-	SampleBuffer(std::vector<sampleFrame> data, int sampleRate);
-	SampleBuffer(
-		const sampleFrame* data, int numFrames, int sampleRate = Engine::audioEngine()->processingSampleRate());
+	enum class Source
+	{
+		Unknown,
+		AudioFile,
+		Base64
+	};
+
+	//! passkey idiom
+	class Access
+	{
+	public:
+		friend class SampleBuffer;
+		friend class SampleLoader;
+		Access(Access&&) = default;
+	private:
+		Access() {}
+		Access(const Access&) = default;
+	};
+
+	SampleBuffer() = delete;
+	SampleBuffer(Access) {}
+	SampleBuffer(Access, const QString& audioFile);
+	SampleBuffer(Access, const QString& base64, int sampleRate);
+	SampleBuffer(Access, std::vector<sampleFrame> data, int sampleRate);
+	SampleBuffer(Access, const sampleFrame* data, int numFrames, int sampleRate);
+
+	static auto create() -> std::shared_ptr<const SampleBuffer>;
+	static auto create(const QString& audioFile) -> std::shared_ptr<const SampleBuffer>;
+	static auto create(const QString& base64, int sampleRate) -> std::shared_ptr<const SampleBuffer>;
+	static auto create(std::vector<sampleFrame> data, int sampleRate) -> std::shared_ptr<const SampleBuffer>;
+	static auto create(const sampleFrame* data, int numFrames,
+		int sampleRate = Engine::audioEngine()->processingSampleRate()) -> std::shared_ptr<const SampleBuffer>;
+
+	~SampleBuffer() = default;
 
 	friend void swap(SampleBuffer& first, SampleBuffer& second) noexcept;
+
+	auto get() const -> std::shared_ptr<const SampleBuffer>;
+
 	auto toBase64() const -> QString;
 
-	auto audioFile() const -> const QString& { return m_audioFile; }
+	auto source() const -> const QString& { return m_source; }
+	auto sourceType() const -> Source { return m_sourceType; }
+	auto audioFileAbsolute() const -> const QString&;
+	auto audioFileRelative() const -> QString; //!< use when saving to project files
+	auto base64() const -> const QString&;
```

Comment:
I just made it a const reference because I didn't want an extra copy if it wasn't needed.
But now I'm rethinking this code, so it will probably change anyway.